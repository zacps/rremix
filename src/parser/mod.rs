use pest_derive::*;

#[derive(Parser)]
#[grammar = "parser/grammar.pest"]
pub struct RemixParser;

#[cfg(test)]
mod tests {
    use crate::format_pair;
    use insta::assert_snapshot;
    use pest::Parser;
    use std::fs;
    use test_case::test_case;

    use super::*;

    #[test_case("../src/standard-lib.rem")]
    //#[test_case("ex/factorial.rem")]
    //#[test_case("ex/factorial2.rem")]
    //#[test_case("ex/primes.rem")]
    #[test_case("ex/test1.rem")]
    #[test_case("ex/test2.rem")]
    #[test_case("ex/test3.rem")]
    #[test_case("ex/test4.rem")]
    #[test_case("ex/test5.rem")]
    #[test_case("ex/test6.rem")]
    #[test_case("ex/test7.rem")]
    #[test_case("ex/test8.rem")]
    #[test_case("ex/test9.rem")]
    #[test_case("ex/test10.rem")]
    #[test_case("ex/test11.rem")]
    #[test_case("ex/test12.rem")]
    #[test_case("ex/test13.rem")]
    #[test_case("ex/test14.rem")]
    #[test_case("ex/test15.rem")]
    #[test_case("ex/test16.rem")]
    #[test_case("ex/test17.rem")]
    #[test_case("ex/test18.rem")]
    #[test_case("ex/test19.rem")]
    #[test_case("ex/test20.rem")]
    #[test_case("ex/test21.rem")]
    #[test_case("ex/test22.rem")]
    #[test_case("ex/test23.rem")]
    #[test_case("ex/test24.rem")]
    #[test_case("ex/test25.rem")]
    #[test_case("ex/test26.rem")]
    // #[test_case("ex/test27.rem")]
    #[test_case("ex/test28.rem")]
    //#[test_case("gr/drawing.rem")]
    //#[test_case("gr/arrows.rem")]
    //#[test_case("gr/squares.rem")]
    //#[test_case("gr/bounce.rem")]
    //#[test_case("ob/obj-demo.rem")]
    //#[test_case("ob/sidewindermaze.rem")]
    //#[test_case("ex/test29.rem")]
    //#[test_case("ex/test30.rem")]
    //#[test_case("ex/test31.rem")]
    //#[test_case("ex/test32.rem")]
    //#[test_case("ex/test33.rem")]
    //#[test_case("ex/test34.rem")]
    //#[test_case("ex/test35.rem")]
    //#[test_case("ex/test36.rem")]
    //#[test_case("ex/test37.rem")]
    //#[test_case("ex/test38.rem")]
    //#[test_case("ex/test39.rem")]
    fn test_parser(path: &'static str) {
        let file =
            fs::read_to_string("tests/".to_owned() + path).expect("could not find test file");
        match RemixParser::parse(Rule::program, &file) {
            Ok(mut parse) => {
                let pair = parse.next();
                assert_snapshot!(path, format_pair(&pair.unwrap(), 0, true))
            }
            Err(e) => {
                eprintln!("{e}");
                panic!("Failed to parse file {path}")
            }
        }
    }
}
