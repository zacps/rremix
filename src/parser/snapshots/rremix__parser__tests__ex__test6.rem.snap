---
source: src/parser/mod.rs
assertion_line: 72
expression: "format_pair(&pair.unwrap(), 0, true)"
---
- program
  - function_definition
    - function_signature
      - name_paren > name_part > name_list > single_name_part: "repeat"
      - name_paren > name_part > name_list > single_name_part: "until"
      - name_paren > paren > single_name_part: "condition-block"
      - name_paren > paren > single_name_part: "block"
    - colon: ":"
    - function_statements
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "do"
        - expression > unary_expression > simple_expression > variable: "block"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "if"
        - expression > unary_expression > simple_expression > function_call
          - single_name_part: "do"
          - expression > unary_expression > simple_expression > variable: "condition-block"
        - newindent_block > statement > expression > unary_expression > simple_expression > function_call
          - single_name_part: "break"
          - single_name_part: "out"
      - statement > redo_statement > redo: "redo"
  - function_definition
    - function_signature
      - name_paren > name_part > name_list > single_name_part: "line"
      - name_paren > paren > single_name_part: "string"
      - name_paren > paren > single_name_part: "value"
    - colon: ":"
    - function_statements
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "show"
        - expression > unary_expression > simple_expression > variable: "string"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "show"
        - expression > unary_expression > simple_expression > variable: "value"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "new"
        - single_name_part: "line"
  - statement > assignment_statement
    - variable: "conditionA"
    - colon: ":"
    - expression > unary_expression > simple_expression > literal > boolean: "false"
  - statement > assignment_statement
    - variable: "conditionB"
    - colon: ":"
    - expression > unary_expression > simple_expression > literal > boolean: "false"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "line"
    - literal > string: "\"conditionA: \""
    - expression > unary_expression > simple_expression > variable: "conditionA"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "line"
    - literal > string: "\"conditionB: \""
    - expression > unary_expression > simple_expression > variable: "conditionB"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "repeat"
    - single_name_part: "until"
    - block > statement > expression > unary_expression > simple_expression > variable: "conditionA"
    - newindent_block
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "line"
        - literal > string: "\"inside conditionA: \""
        - expression > unary_expression > simple_expression > variable: "conditionA"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "line"
        - literal > string: "\"inside conditionB: \""
        - expression > unary_expression > simple_expression > variable: "conditionB"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "if"
        - expression > unary_expression > simple_expression > variable: "conditionB"
        - newindent_block > statement > assignment_statement
          - variable: "conditionA"
          - colon: ":"
          - expression > unary_expression > simple_expression > literal > boolean: "true"
      - statement > assignment_statement
        - variable: "conditionB"
        - colon: ":"
        - expression > unary_expression > simple_expression > literal > boolean: "true"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "line"
    - literal > string: "\"conditionA: \""
    - expression > unary_expression > simple_expression > variable: "conditionA"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "line"
    - literal > string: "\"conditionB: \""
    - expression > unary_expression > simple_expression > variable: "conditionB"
  - EOI: ""
